### ISSUE TITLE
[Bug] JS/TS `setCompilerOptions` and `setDiagnosticOptions` only work during initialization

### ISSUE THREAD

**Issue by spahnke** [microsoft/monaco-editor#2841]
### Reproducible in vscode.dev or in VS Code Desktop?

- [X] Not reproducible in [vscode.dev](https://vscode.dev) or VS Code Desktop

### Reproducible in the monaco editor playground?

- [ ] Not reproducible in [the monaco editor playground](https://microsoft.github.io/monaco-editor/playground.html)

### Monaco Editor Playground Code

```typescript
monaco.languages.typescript.javascriptDefaults.setDiagnosticsOptions({
    noSemanticValidation: false,
    noSyntaxValidation: false,
    diagnosticCodesToIgnore: []
});

monaco.languages.typescript.javascriptDefaults.setCompilerOptions({
    target: monaco.languages.typescript.ScriptTarget.ESNext,
    lib: ["esnext"],
    allowNonTsExtensions: true
});

const editor = monaco.editor.create(document.getElementById("container"), {
    value: `return 1;`,
    language: "javascript"
});

let topLevelReturn = false;
editor.addAction({
    id: "toggle_top_level_return",
    label: "Toggle Top Level Return",
    keybindings: [monaco.KeyMod.chord(monaco.KeyMod.CtrlCmd | monaco.KeyCode.KeyD, monaco.KeyMod.CtrlCmd | monaco.KeyCode.KeyR)],
    contextMenuGroupId: "7_test",
    run: () => {
        if (topLevelReturn) {
            monaco.languages.typescript.javascriptDefaults.setDiagnosticsOptions({
                noSemanticValidation: false,
                noSyntaxValidation: false,
                diagnosticCodesToIgnore: []
            });
            topLevelReturn = false;
        } else {
            monaco.languages.typescript.javascriptDefaults.setDiagnosticsOptions({
                noSemanticValidation: false,
                noSyntaxValidation: false,
                diagnosticCodesToIgnore: [1108]
            });
            topLevelReturn = true;
        }
    }
});

let browserCompletion = false;
editor.addAction({
    id: "toggle_browser_completion",
    label: "Toggle Browser Completion",
    keybindings: [monaco.KeyMod.chord(monaco.KeyMod.CtrlCmd | monaco.KeyCode.KeyD, monaco.KeyCode.KeyB)],
    contextMenuGroupId: "7_test",
    run: () => {
        if (browserCompletion) {
            monaco.languages.typescript.javascriptDefaults.setCompilerOptions({
                target: monaco.languages.typescript.ScriptTarget.ESNext,
                lib: ["esnext"],
                allowNonTsExtensions: true
            });
            browserCompletion = false;
        } else {
            monaco.languages.typescript.javascriptDefaults.setCompilerOptions({
                target: monaco.languages.typescript.ScriptTarget.ESNext,
                lib: ["esnext", "dom"],
                allowNonTsExtensions: true
            });
            browserCompletion = true;
        }
    }
});
```

### Actual Behavior

You can't toggle between DOM completions vs. just JS completions, and top-level return vs. no top-level return (see respective actions).

### Expected Behavior

Using the actions toggles the respective behavior on and off.

### Additional Context

I have been using `setCompilerOptions` and `setDiagnosticOptions` to toggle between DOM completions vs. just JS completions, and top-level return vs. no top-level return. This seems to be broken with version 0.31.1. The options work initially if I set the corresponding values, but not later (simulated here trough an action). It seems as if the workers don't refresh anymore. Is there something I need to do now to accomplish that that I've missed from the changelog or is this a bug?

**Comment by spahnke** [comment]
I tried to debug this a bit today but ran into a wall. The cause of this seems to be a `_listeners` property that is `undefined` on the `_onDidChange` event emitter when called here from the editor action/timeout.

https://github.com/microsoft/monaco-editor/blob/9af034d1cb50361bd5237ca79d00b62f2b5ef4ac/src/typescript/monaco.contribution.ts#L629

The funny thing is, when the subscribers subscribe to that event here
https://github.com/microsoft/monaco-editor/blob/9af034d1cb50361bd5237ca79d00b62f2b5ef4ac/src/typescript/workerManager.ts#L25
and here 
https://github.com/microsoft/monaco-editor/blob/9af034d1cb50361bd5237ca79d00b62f2b5ef4ac/src/typescript/languageFeatures.ts#L275
the `_listeners` object gets initialized/is present.

I marked the `LanguageServiceDefaultsImpl` instance and the event emitter/IEvent instance with a simple counter to make sure it is the same instance. That seems to be the case. But still, somehow the `_listeners` property is `undefined` when the event is fired after a configuration change 😕 

```ts
let fooCounter = 0;
class LanguageServiceDefaultsImpl implements LanguageServiceDefaults {
	private _onDidChange = new Emitter<void>();
	// ...
	constructor(
		compilerOptions: CompilerOptions,
		diagnosticsOptions: DiagnosticsOptions,
		workerOptions: WorkerOptions,
		inlayHintsOptions: InlayHintsOptions
	) {
		//...
		// @ts-ignore
		this._onDidChange.foo = ++fooCounter; this.onDidChange.foo = fooCounter; this.foo = fooCounter;
                // ...
```
